lp_solve/lp_solve.1
.TH LP_SOLVE 1ES
.SH NAME
lp_solve	- Solve (mixed integer) linear programming problem.
.SH SYNOPSIS
lp_solve [option]* "<" <input-file>
.SH OPTIONS
.TP 1.2i
-v
Verbose mode. Among other things, shows all the pivots.
.TP
-h
Help mode, prints the usage.
.TP
-d
Debug mode, all intermediate results are printed, and the branch-and-bound
decisions in case of (mixed) integer problems.
.TP
-p
Only functional for pure LP problems. Print the values of the dual
variables as well in the result. They are named r_1 until r_XXXXX unless
specified by the user.  Note that bounds (constraints on just one variable)
are not considered real constraints, and are not given a row in the matrix,
and are therefore not printed here.
.TP
-b <bound>
Specify an upper (when minimizing) or lower (when maximizing) limit for the
value of the objective function to
the program. Only useful for (mixed) integer problems.  If close enough, may
speed up the calculations. The same result can be obtained by adding an extra
constraint to the problem.
.TP
-c
When branching in MILP problems, take the ceiling of the selected non-integer
variable first instead of the floor. This can influence the speed of MILP
problems.
.TP
-e <value>
Specify the accuracy with which it is checked whether the value of a variable
is really integer. <value> must be between 0 and 0.5. Default value is 1e-6
and should be OK for most applications. Of course only useful for MILP
problems.
.TP
-i
Print all intermediate valid solutions. Can give you useful
solutions even if the total run time is too long.
Only useful for (mixed) integer problems.
.TP
-s
Both rows and columns are scaled according to the geometric mean of the
coefficients on them before solving. This might improve the numerical
stability of your problem.
.TP
-I
Print info after reinverting.
.TP
-t
Trace pivot selection.
.TP
-mps
Read from MPS file instead of lp file.
.TP
-degen
Use random perturbations to reduce degeneracy, can increase numerical
instability.
.SH DESCRIPTION
The linear programming problem can be formulated as: Solve A.x >= V1, with
V2.x maximal. A is a matrix, x a vector of (nonnegative) variables, V1 a
vector called the right hand side, and V2 a vector specifying the objective
function.
.br
Any number of the variables may be specified to be of type integer.
.br
This program solves problems of this kind. It is slightly more general than
the above problem, in that every row of A (specifying one constraint) can have
its own (in)equality, <=, >= or =. The result specifies values for all
variables.
.br
Uses a 'Simplex' algorithm and sparse matrix methods, for pure LP problems.
If one or more of the variables is declared integer, the Simplex algorithm is
iterated with a branch and bound algorithm, until the desired optimal
solution is found.
.SH "INTERRUPT"
In order to make it possible to get useful solutions for a mixed integer
problem which takes too long to finish completely, the "interrupt" signal
(kill -INT, ^C for many people) will not terminate the program when it is
solving a MILP problem, but rather print the best valid solution found so
far. This feature is not useful and thus disabled when you are solving a pure
LP problem. If you see all zeros, it implies that no valid solution has been
found yet. To really terminate the program, use a different signal, like KILL.
.br
The "-i" option will print all intermediate valid solutions.
.SH "INPUT SYNTAX"
The default input syntax is a set of algebraic expressions and "int"
declarations in the following order:
.sp
<objective function>
.br
<constraint>+
.br
<declaration>*
.sp
where:
.TP 0.2i
-
<objective function> is a linear combination of variables, ending with a
semicolon, optionally preceded by "max: " or "min: " to indicate whether you
want it to be minimized or maximized. The case is not important, "Max:" or
"MAX:" will work as well. Maximization is the default.
.TP
-
<constraint> is an optional constraint name followed by a colon plus a
linear combination of variables and constants, followed by a relational
operator, followed again by a linear combination of variables and constants,
ending with a semicolon. The relational operator can be any of the following:
"<" "<=" "=" ">" ">=". There is no semantic difference between "<" and "<="
nor between ">" and ">=" (even for integer variables!).
.TP
-
<declaration> is of the form: "int" <var>+ ";" Commas are allowed between
variables.
.sp
So, the simplest linear problem consists of an objective function and 1
constraint.
.SH EXAMPLE
The simple problem:
.sp
x1 >= 1
.br
x2 >= 1
.br
x1 + x2 >= 2
.br
minimize x1 + x2 (= maximize -(x1 + x2)), with x1 integer
.sp
can be written as follows:
.sp
-x1 + -x2;
.br
(or min: x1 + x2;)
.br
x1 > 1;
.br
x2 > 1;
.br
x1 + x2 > 2;
.br
int x1;
.sp
The correct result for (x1, x2) is of course (1, 1).
.br
With the -mps option, lp_solve will accept MPS as input format.
.SH BUGS
Specifying a constraint name for a bound (constraints on just single
variables) does not have an effect: they are not stored inside the main matrix
and are not assigned a dual variable.
.TP
-
The problem consists entirely of constraints on just single variables
(so-called "bounds", like x < 1; ) and no constraint with more than 1
variable (like x + 3 y > 17; ). This leaves lp_solve with an empty problem
matrix, as bounds are not stored in the main matrix. No real-life examples
should be of this form, so I am not really chasing this problem.
.TP
-
Many people forget that lp_solve can only handle POSITIVE values for the
variables. While reading MPS files it will however handle free or negative
variables by replacing them with a variable pair <var>_neg and <var>_pos or
-<var> respectively. It is up to the user to interpret the result of this
transformation.
.TP
- Sometimes problems are numerically unstable, and the unavoidable rounding
errors inside lp_solve will cause aborts. It is very hard to give general
solutions to this problem, but try to keep all values in your problem in the
order of magnitude of 1 by proper scaling. This is almost always better than
using lp_solves built-in scaling (with -s). Almost parallel constraints are
also not very good for numerical stability. Use "lp_solve -v" and observe the
values of the pivots to see if there are any dangerously large or low numbers
there.
.br
Building lp_solve with long doubles (see the Makefile) can help to increase
numerical stability, but will also increase the run time considerably.
.br
You can consult the author as well if you encounter numerical problems, but
please remember that it is very easy to formulate an infeasible LP problem, so
be sure there is a solution.
.SH SEE ALSO
The implementation of the simplex kernel was mainly based on:
.br
W. Orchard-Hays: "Advanced Linear Programming Computing Techniques",
McGraw-Hill 1968
.br
The mixed integer branch and bound part was inspired by:
.br
section 6.4 of "An Introduction to Linear Programming and Game Theory" by
Paul R. Thie, second edition published by John Wiley and Sons in 1988.
.br
This book refers to:
.br
Dakin, R.J., "A Tree Search Algorithm for MILP Problems", Comput. J., 8 (1965)
pp. 250-255
.SH ACKNOWLEDGEMENTS
The work of Jeroen Dirks made the transition from the basic version 1.5 to
the full version 2.0 possible. He contributed the procedural interface, a
built-in MPS reader, and many fixes and enhancements to the code.
.SH CONTRIBUTED BY
M.R.C.M. Berkelaar
.br
Eindhoven University of Technology
.br
Design Automation Section
.br
P.O. Box 513
.br
NL-5600 MB Eindhoven, The Netherlands
.br
phone ...-31-40-473345
.br
E-mail: michel@es.ele.tue.nl
.SH STATUS
Use at own risk. Bug reports are welcome, as well as success stories.


NEWS
NEWS for Mercury 22.01.9
========================

This is a bug-fix release.

* [Github issue #147]. We have fixed a const correctness issue in the C
  runtime that was causing warnings to be emitted with recent versions of
  GCC and the GNU C library.

* [Mantis bug #566]. We have fixed an issue that was causing GNU Make 4.4.0
  and above to emit warnings about undefined variables.

* We have improved the compatibility of the `configure` script with older
  versions of `bash`.

* We have fixed a bug in the C# implementation of
  `random.system_rng.generate_uint8/4` that caused it to throw an exception
  after successfully generating a random value.

* We have added the missing lower-bound checks on the time components of the
  predicate `calendar.init_date/8`.

* We have fixed two bugs in `calendar.duration_from_string/2` that caused it
  to accept invalid duration strings. The first was that the string "P" was
  incorrectly being accepted as the zero duration; the second was that duration
  strings containing a fractional second component with no whole number part
  (e.g. "PT.5S") were incorrectly being accepted.

* We have fixed a bug in `calendar.duration/2` that caused it to compute
  an incorrect duration, or abort, for some pairs of dates whose difference
  required borrowing days across months of differing length.

* `calendar.date_from_string/2` and `calendar.duration_from_string/2` now
   check for integer overflow in their components.

* We have fixed bugs in the C# and Java implementations of
  `math.unchecked_log2/1` that caused target language compilation errors
  when compiling with `--intermodule-optimisation`.

* We have fixed a bug where using C# contextual keywords as the names
  of Mercury entities (e.g. module names) could cause invalid C# to be
  generated.

* We have fixed a bug where a Mercury type containing a data constructor whose
  name begins with a digit could cause invalid C# code to be generated.

* All remaining Java security manager dependencies have been replaced,
  since the security manager is deprecated.

* We have fixed an issue that caused the Microsoft linker to emit
  warning `LNK4001`: "no object files specified; libraries used".

* We have fixed an incorrect overflow check in `string.to_int/2`,
  `string.to_uint`, `string.base_string_to_int/3` and
  `string.base_string_to_uint/3`.

* We have fixed a bug where `string.to_uint/2` and
  `string.base_string_to_uint/3` returned zero instead of failing
  when called with an empty input string.

* Many documentation fixes and improvements.

NEWS for Mercury 22.01.8
========================

This is a bug-fix release.

* [Mantis bug #565]. We have reverted a change that broke `mmake` in Mercury
  versions 22.01.6 and 22.01.7.

* We have fixed a problem that was causing the generation of the debugger
  documentation to fail on Cygwin.

* We have increased the C stack size for the compiler on 64-bit Cygwin systems,
  due to stack overflows occuring with the default size.

* We have fixed a bug in the debugger, which, in some rare circumstances, could
  cause it to print terms to the wrong stream.

NEWS for Mercury 22.01.7
========================

This is a bug-fix release.

* [Github issue #123]. We have fixed a problem that was causing compilation
  with GCC 13 to fail.

* We have renamed a macro used in the compiler to avoid a clash with
  the system headers in the MSYS2 UCRT64 environment.

* We have fixed a problem that was causing the generation of the debugger
  documentation to fail on Windows.

* We have fixed a problem where MSVC could not compile the C code generated for
  some tabled Mercury predicates. This was caused by the C code inadvertently
  making use of some GNU C extensions that MSVC does not support.

NEWS for Mercury 22.01.6
========================

This is a bug-fix release.

* We have fixed the implementations of transitive closure and reflexive
  transitive closure in the `digraph` module.

* We have a fixed a bug in `string.format` and related predicates where `int8`,
  `int16` and `int32` values were being incorrectly promoted to `int` and sign
  extended when one of the unsigned conversion specifiers was applied.

* [Github issue #116]. We have fixed a problem building programs with
  `--intermodule-optimization` where the compiler was searching for library
  files that were not installed unless the library was also built with
  `--intermodule-optimization` enabled.

* [Github issue #117]. We have fixed a bug where the compiler would silently
  fail when compiling a program that consists of a single source file and
  some of the required libraries are not available in the selected grade.

* The documentation of the following predicates in the `term_io` module has
  been fixed:

    - pred `quote_char/3`
    - pred `quote_atom/3`
    - pred `write_escaped_char/3`
    - pred `write_escaped_string/3`

  Their documentation said that they write to `stdout`, when in fact
  they write to the current text output stream. By default, this is `stdout`,
  but calls to `io.set_output_stream` may change this.

NEWS for Mercury 22.01.5
========================

This is a bug-fix release.

* We have fixed a bug that was causing invalid fact table file names
  to be written to dependency files.

* We have fixed a bug with the MLDS backend that was causing
  crashes with certain switches on strings.

* We have fixed a bug that was causing some transitive intermodule
  optimization interfaces to be silently ignored.

* We have fixed a bug where termination analysis could cause a compiler abort
  when invoked on some rarely-occurring predicate structures.

NEWS for Mercury 22.01.4
========================

This is a bug-fix release.

* We have fixed and clarified the behaviour of the following `string`
  module predicates when called on a string containing ill-formed code
  unit sequences:

    - pred `all_match/2`
    - pred `index_next_repl/5`
    - pred `unsafe_index_next_repl/5`

lp_solve/lp_solve.c
#include <string.h>
#include <stdio.h>
#include "lpkit.h"
#include "lpglob.h"
#include "patchlevel.h"

void print_help(char *argv[])
{
  printf("Usage of %s version %s:\n", argv[0], PATCHLEVEL);
  printf("%s [options] \"<\" <input_file>\n", argv[0]);
  printf("list of options:\n");
  printf("-h\t\tprints this message\n");
  printf("-v\t\tverbose mode, gives flow through the program\n");
  printf("-d\t\tdebug mode, all intermediate results are printed,\n\t\tand the branch-and-bound decisions\n");
  printf("-p\t\tprint the values of the dual variables\n");
  printf("-b <bound>\tspecify a lower bound for the objective function\n\t\tto the program. If close enough, may speed up the\n\t\tcalculations.\n");
  printf("-i\t\tprint all intermediate valid solutions.\n\t\tCan give you useful solutions even if the total run time\n\t\tis too long\n");
  printf("-e <number>\tspecifies the epsilon which is used to determine whether a\n\t\tfloating point number is in fact an integer.\n\t\tShould be < 0.5\n");
  printf("-c\t\tduring branch-and-bound, take the ceiling branch first\n");
  printf("-s\t\tuse automatic problem scaling.\n");  
  printf("-I\t\tprint info after reinverting\n");
  printf("-t\t\ttrace pivot selection\n");
  printf("-mps\t\tread from MPS file instead of lp file\n");
  printf("-degen\t\tuse perturbations to reduce degeneracy,\n\t\tcan increase numerical instability\n");
}

int main(int argc, char *argv[])
{
  lprec *lp;
  int i;
  short verbose = FALSE;
  short debug = FALSE;
  short print_sol = FALSE;
  short print_duals = FALSE;
  short floor_first = TRUE;
  short scaling = FALSE;
  short print_at_invert = FALSE;
  short tracing = FALSE;
  short mps = FALSE;
  short anti_degen = FALSE;
  int result;
  REAL obj_bound = (REAL)DEF_INFINITE;
  REAL epsilon = (REAL)DEF_EPSILON;

  for(i = 1; i < argc; i++)
    {
      if(strcmp(argv[i], "-v") == 0)
	verbose = TRUE;
      else if(strcmp(argv[i], "-d") == 0)
	debug = TRUE;
      else if(strcmp(argv[i], "-i") == 0)
	print_sol = TRUE;
      else if(strcmp(argv[i], "-c") == 0)
	floor_first = FALSE;
      else if(strcmp(argv[i], "-b") == 0)
	obj_bound = atof(argv[++i]);
      else if(strcmp(argv[i], "-e") == 0)
	{
	  epsilon = atof(argv[++i]);
	  if((epsilon <= 0.0) || (epsilon >= 0.5))
	    {
	      fprintf(stderr, "Invalid epsilon %g; 0 < epsilon < 0.5\n",
		      (double)epsilon);
	      exit(1);
	    }
	}
      else if(strcmp(argv[i], "-p") == 0)
	print_duals = TRUE;
      else if(strcmp(argv[i], "-h") == 0)
        {
           print_help(argv); 
           exit(0);
        }
      else if(strcmp(argv[i], "-s") == 0)
	scaling = TRUE;
      else if(strcmp(argv[i], "-I") == 0)
        print_at_invert=TRUE;
      else if(strcmp(argv[i], "-t") == 0)
        tracing=TRUE;
      else if(strcmp(argv[i],"-mps") == 0)
        mps=TRUE;
      else if(strcmp(argv[i],"-degen") == 0)
        anti_degen=TRUE;
    }

  if(mps)
    lp = read_mps(stdin, verbose);
  else
    lp = read_lp_file(stdin, verbose, "lp" );

  if(lp->columns < 8 && verbose)
    print_lp(lp);
  if(scaling)
    auto_scale(lp);
  if(print_sol)
    lp->print_sol = TRUE;
  lp->epsilon = epsilon;
  if(print_duals)
    lp->print_duals = TRUE;
  if(debug)
    lp->debug = TRUE;
  if(!floor_first)
    lp->floor_first = FALSE;
  if(print_at_invert)
    lp->print_at_invert = TRUE;
  if(tracing == TRUE)
    lp->trace = TRUE;
  if(obj_bound != DEF_INFINITE)
    lp->obj_bound = obj_bound;
  lp->anti_degen = anti_degen;

  if(verbose)
    {
       printf("Solving\n");
       lp->verbose = TRUE;
    }

  result = solve(lp);

  if(result == OPTIMAL)
    {
       print_solution(lp);
       if(verbose)
         fprintf(stderr,
		 "Branch & Bound depth: %d\nNodes processed: %d\nSimplex pivots: %d\n",
		 lp->max_level, lp->total_nodes, lp->total_iter);
    }

  if(result == INFEASIBLE)
    printf("This problem is infeasible\n");
  if(result == UNBOUNDED)
    printf("This problem is unbounded\n");
  if(result == FAILURE)
    printf("lp_solve failed\n");
   return(result);
}
